### ISSUE TITLE
GeneSplicer plugin not working when running in google cloud 

### ISSUE THREAD

**Issue by MatteoSchiavinato** [Ensembl/ensembl-vep#1590]
Hi, 

I have recently migrated a nextflow pipeline I am using from local to google cloud. This pipeline uses VEP to annotate variants among other things. I have moved to a google cloud storage bucket all the annotation sources that I use with VEP, and I have a docker image with VEP executable and its plugins. 

This configuration (databases + image + nextflow pipe) has worked perfectly in local until now, but ever since I migrated it to google cloud, I have started getting this error at the stage where GeneSplicer is used within VEP: 

```
Use of uninitialized value $test in pattern match (m//) at /opt/vep/Plugins/GeneSplicer.pm line 122.
Use of uninitialized value $test in concatenation (.) or string at /opt/vep/Plugins/GeneSplicer.pm line 122.
WARNING: Failed to instantiate plugin GeneSplicer: ERROR: failed to run genesplicer binary:

[W::hts_idx_load2] The index file is older than the data file: clinvar.GRCh37.vcf.gz.tbi
[W::hts_idx_load2] The index file is older than the data file: ExAC.0.3.GRCh37.vcf.gz.tbi
[W::hts_idx_load2] The index file is older than the data file: gnomad.exomes.r2.1.1.vcf.gz.tbi
[W::hts_idx_load2] The index file is older than the data file: gnomad.genomes.r2.1.1.vcf.gz.tbi
[W::hts_idx_load2] The index file is older than the data file: ESP6500SI-V2-SSA137.vcf.gz.tbi
[W::hts_idx_load2] The index file is older than the data file: GCF_000001405.25.bgz.tbi
```

This is the full VEP command: 
```
/opt/vep/src/ensembl-vep/vep \
--offline --cache --force_overwrite \
--show_ref_allele \
--numbers \
--fork 64 \
--refseq \
--cache_version 104 \
--dir_cache cache_vep \
--fasta vep_zipped_fasta/Homo_sapiens.GRCh37.75.dna.primary_assembly.fa.gz \
--dir_plugins /opt/vep/Plugins \
--assembly GRCh37 \
--custom clinvar.GRCh37.vcf.gz,ClinVar,vcf,exact,0,CLNSIG,CLNREVSTAT,CLNDN,CLNDISDB,CLNALLELEID  \
--custom ExAC.0.3.GRCh37.vcf.gz,exac03,vcf,exact,0,AF,AC,AC_Het,AC_Hom,AC_AFR,AC_AMR,AC_EAS,AC_FIN,AC_NFE,AC_OTH,AC_SAS,AN_AFR,AN_AMR,AN_EAS,AN_FIN,AN_NFE,AN_OTH,AN_SAS,Het_AFR,Het_AMR,Het_EAS,Het_FIN,Het_NFE,Het_OTH,Het_SAS,Hom_AFR,Hom_AMR,Hom_EAS,Hom_FIN,Hom_NFE,Hom_OTH,Hom_SAS  \
--custom gnomad.exomes.r2.1.1.vcf.gz,gnomad_exomes,vcf,exact,0,AF,AF_afr,AF_amr,AF_asj,AF_eas,AF_fin,AF_nfe,AF_oth,AF_sas,AF_XY,AF_XX,popmax \
--custom gnomad.genomes.r2.1.1.vcf.gz,gnomad_genomes,vcf,exact,0,AF,AF_afr,AF_amr,AF_asj,AF_eas,AF_fin,AF_nfe,AF_oth,AF_sas,AF_XY,AF_XX,popmax \
--custom ESP6500SI-V2-SSA137.vcf.gz,esp6500siv2,vcf,exact,0,MAF \
--custom GCF_000001405.25.bgz,avsnp152,vcf,exact,0,RS \
--custom hg19_genomicSuperDups_name.bed.gz,hg19_genomicSuperDups_name,bed,overlap,0 \
--custom hg19_genomicSuperDups_fracMatch.bed.gz,hg19_genomicSuperDups_fracMatch,bed,overlap,0 \
--custom hg19_genomicSuperDups_fracMatchIndel.bed.gz,hg19_genomicSuperDups_fracMatchIndel,bed,overlap,0  \
--custom hg19_icgc28.vcf.gz,icgc28,vcf,exact,0,OCCURRENCE  \
--custom hg19_nci60.vcf.gz,nci60,vcf,exact,0,AF \
--plugin LOVD \
--plugin dbscSNV,dbscSNV1.1_GRCh37.txt.gz \
--plugin dbNSFP,dbNSFP4.3a_GRCh37.gz,transcript_match=1,ALL \
--plugin satMutMPRA,file=satMutMPRA_GRCh37_ALL.gz,cols=ALL \
--plugin GeneSplicer,genesplicer_path/sources/genesplicer,genesplicer_path/human \
--plugin MaxEntScan,maxentscan_path/,NCSS \
--plugin SpliceAI,snv=spliceai_scores.raw.snv.hg19.vcf.gz,indel=spliceai_scores.raw.indel.hg19.vcf.gz \
--plugin pLI,plI_gene.txt \
--plugin NearestExonJB \
--af_1kg \
--hgvs --hgvsg --symbol --nearest symbol --distance 1000 \
--canonical \
--exclude_predicted \
--regulatory \
-i HG005_NIST_chr22.decomposed.normalized.span.vcf \
-o HG005_NIST_chr22_VEP_output.raw.vcf \
-vcf --flag_pick_allele_gene  \
--pick_order mane,canonical,appris,tsl,biotype,ccds,rank,length
```

This is my system's properties: 
- VEP version: 107
- VEP Cache version: 107
- Perl version: v5.32.1
- OS: Ubuntu 
- tabix installed: yes 

Could you please let me know if there is something I should know regarding vep's usage in cloud?

**Comment by olaaustine** [comment]
Hi @MatteoSchiavinato, 
Hope this meets you well? 
Thank you for your query. 
Can you please confirm that the GeneSplicer binaries provided have been compiled?
Thank you 
Ola.

**Comment by MatteoSchiavinato** [comment]
Hi, yes, we've been using it for years.

**Comment by olaaustine** [comment]
Hi @MatteoSchiavinato, 
Thank you for your quick response. 
We suspect that there is an issue with the GeneSplicer binaries.
Did you get outputs for the other plugins? 
Thank you 
Ola.

**Comment by MatteoSchiavinato** [comment]
Yes, we are using other plugins and they have produced an output. We are using a GeneSplicer plugin path that is entirely available in google cloud (a bucket) and viewable with `gsutil ls`. The same folder organisation works locally, so it feels as if GeneSplicer cannot find its own files and create the `$test` variable when trying to read a bucket.

**Comment by MatteoSchiavinato** [comment]
Hi, any news on this matter? We have done some tests but the problem stays. We can confirm that the problem arises only when running the plugin in the cloud (google).

**Comment by olaaustine** [comment]
Hi @MatteoSchiavinato, 
Hope you are well? 
The GeneSplicer files are executable files and others that require a Unix system to work, this might be the issue except if the bucket has been mounted as a file system you have might have issues running the executable in GeneSplicer files. 
Please can you confirm that this has been done? 
Thank you 
Ola.

**Comment by MatteoSchiavinato** [comment]
Thanks for your quick reply. I'm not sure I understand correctly. In order for these files to work, the bucket should be mounted as filesystem?

**Comment by olaaustine** [comment]
Hi @MatteoSchiavinato, 
Just to get better clarification, in your first message, you said you migrated your nextflow pipeline to google cloud, is it possible to share your nextflow.config file? 
 I am trying to debug the issue. 
Thank you 
Ola

**Comment by MatteoSchiavinato** [comment]
Hi, the pipeline is local and the google cloud configuration is done through nextflow. The pipeline has 18 steps, the first 11 work seamlessly and the issue is only genesplicer.

I can share the config file but I wanted to specify that the configuration of google-lifesciences is probably right since the other processes work.

**Comment by MatteoSchiavinato** [comment]
This is the part of the config file where the google cloud profile is defined: 

```
    gcp {

        workDir = "gs://gino-nf-data/nf-work"

        conda {
            enabled=true
        }

        docker {
            enabled=true 
            fixOwnership=true 
            remove=true 
        }
        
        google {
            region='europe-west2'
            location='europe-west2'
            project='geno-296610'
            lifeSciences.bootDiskSize=500.GB
        }

        process {
            executor="google-lifesciences"
            container="eu.gcr.io/geno-296610/gino_nf:3.2.0"

            withLabel: max_threads {
                machineType = 'e2-highcpu-32'
                cpus = 32
                maxForks = 1
            }

            withLabel: mid_threads {
                machineType = 'e2-highcpu-32'
                cpus = 32
                maxForks = 2
            }

            withLabel: intermediate {
                machineType = 'e2-highcpu-32'
                cpus = 8
                maxForks = 4
            }

            withLabel: mid_forks {
                machineType = 'e2-standard-8'
                cpus = 2
                maxForks = 4
            }

            withLabel: max_forks {
                machineType = 'e2-standard-8'
                cpus = 1
                maxForks = 8
            }

            withLabel: max_mem {
                machineType = 'e2-highmem-16'
                cpus = 8
                maxForks = 2
            }
        }

        env {
            AGENT_SCRIPT="/gino/AGeNT/agent/agent.sh"
            AWK="awk"
            BAMCOVERAGE="bamCoverage"
            BCFTOOLS="bcftools"
            BEDTOOLS="bedtools"
            BGZIP="bgzip"
            CNVKIT="cnvkit.py"
            DATA_DIR="/gino/data"
            EXPANSION_HUNTER="/opt/ExpansionHunter/ExpansionHunter-v4.0.2-linux_x86_64/bin/ExpansionHunter"
            FUNCTIONS="/gino/data/functions.json"
            GAWK="gawk"
            GATK="/opt/gatk/gatk-4.3.0.0/gatk"
            PYTHON3="python3"
            RSCRIPT="Rscript"
            SAMBAMBA="/sambamba-0.7.1/sambamba"
            SAMTOOLS="samtools"
            SCRIPTS_DIR="/gino/scripts"
            SENTIEON_INSTALL_DIR="/opt/sentieon/sentieon-genomics-202112.07"
            SENTIEON_LICENSE="10.132.0.5:8990"
            STAR="/opt/star/STAR"
            STARLONG="opt/star/STARlong"
            TABIX="tabix"
            TAPES="/gino/tapes/tapes.py"
            VCFSORT="vcf-sort"
            VEP="/opt/vep/src/ensembl-vep/vep"
            VEP_PLUGINS="/opt/vep/Plugins"
            VT="vt"
            WHATSHAP_CONDA_ENV="/root/miniconda3/etc/profile.d/conda.sh"
        }
    }
```

**Comment by olaaustine** [comment]
Hi @MatteoSchiavinato, 
Thank you for your response. 
Thank you for your patience while we debug this.
Just to confirm can you run a test using the compiled GeneSplicer you have on the cloud, for us to see if there is an output.
`genesplicer <fasta-file> [options]
-f file_name     Write the results in file_name
-d dist          The maximum score within dist bp is chosen (default: 30bp)`
Please let us know if there is an output. 
Thank you 
Ola.

**Comment by MatteoSchiavinato** [comment]
```
gsutil cp gs://ref-database/snp/VEP/GeneSplicer_files/GeneSplicer/sources/genesplicer . 

chmod +x genesplicer 

./genesplicer -h
USAGE:  ./genesplicer <fasta-file> <specific-genome-training-directory> [options] 
Options:
-f file_name     Write the results in file_name
-a t             Choose t as a threshold for the acceptor sites
-d t             Choose t as a threshold for the donor sites
-e n             The maximum acceptor score within n bp is chosen
-i n             The maximum donor score within n bp is chosen
-h               Display the options of the program

```

**Comment by MatteoSchiavinato** [comment]
This is the code line where I have it running as a VEP plugin: 
```
--plugin GeneSplicer,${GeneSplicer}/sources/genesplicer,${GeneSplicer}/human
```
Where `${GeneSplicer}` is a variable containing the following path: 
```
gs://ref-database/snp/VEP/GeneSplicer_files/GeneSplicer
```
And `${GeneSplicer}` path contains the following subdirs:
```
gsutil ls gs://ref-database/snp/VEP/GeneSplicer_files/GeneSplicer

gs://ref-database/snp/VEP/GeneSplicer_files/GeneSplicer/COPYRIGHT
gs://ref-database/snp/VEP/GeneSplicer_files/GeneSplicer/LICENSE
gs://ref-database/snp/VEP/GeneSplicer_files/GeneSplicer/README
gs://ref-database/snp/VEP/GeneSplicer_files/GeneSplicer/arabidopsis/
gs://ref-database/snp/VEP/GeneSplicer_files/GeneSplicer/bin/
gs://ref-database/snp/VEP/GeneSplicer_files/GeneSplicer/drosophila/
gs://ref-database/snp/VEP/GeneSplicer_files/GeneSplicer/human/
gs://ref-database/snp/VEP/GeneSplicer_files/GeneSplicer/malaria/
gs://ref-database/snp/VEP/GeneSplicer_files/GeneSplicer/pyoelii/
gs://ref-database/snp/VEP/GeneSplicer_files/GeneSplicer/rice/
gs://ref-database/snp/VEP/GeneSplicer_files/GeneSplicer/sources/
gs://ref-database/snp/VEP/GeneSplicer_files/GeneSplicer/training_data_sets/
```

Note that `human/` is there, which is the one we pass as argument.

**Comment by MatteoSchiavinato** [comment]
The plugin is correctly installed in a path named `/opt/vep/Plugins` which is inside a docker image. We do pass the `--dir_plugins` argument to VEP, and as mentioned before, locally it works fine. The issue is just in the cloud. 

Thanks again for helping out!

**Comment by MatteoSchiavinato** [comment]
Lastly: it isn't the only plugin that we pass, the others however work also in the cloud. The issue seems related to genesplicer only.

**Comment by MatteoSchiavinato** [comment]
Hi, I have new insights into this. It may be a very easy fix on your side. 

The `GeneSplicer.pm` plugin script has the following lines:

```
  my $bin = shift @$params;
  die("ERROR: genesplicer binary not found\n") unless -e $bin;
  my $test = `$bin 2>&1`;
  die("ERROR: failed to run genesplicer binary:\n$test\n") unless $test =~ /^USAGE/;
  $self->{_bin} = $bin;
```

These lines fail when the executable is in a google cloud bucket, because `-e` looks for a standard path (`/path/to/...`). 

I have tried to comment out the line where `-e` is run and the command works, because google cloud computing stages the file in the local VM instance and then executes it. Hence, the issue is just Perl trying to assess its existence without being able to parse a `gs://...` path.

**Comment by olaaustine** [comment]
Hi @MatteoSchiavinato, 
Thank you for your patience with this. 
We have added the GeneSplicer binaries to our VEP Docker image and this will be available from release/112. This enhancement ensures that the GeneSplicer binaries are available within the Docker environment. 
We have also updated the GeneSplicer VEP plugin for release/112 to run without explicitly stating the path of the executable file when using Docker. 
Thank you very much. 
Ola.
